src/completion.ml

```diff
@@ -0,0 +1,147 @@
+  let rec build_prod k p =
+    if k = 0 then p
+    else
+      let k = k-1 in
+      let mk_typ = Bindlib.unbox (build_prod k _Type) in
+      let mk = fresh_meta mk_typ k in
+      let tk = _Meta mk (Array.map _Vari (Array.sub vs 0 k)) in
+      let b = Bindlib.bind_var vs.(k) p in
+      let p = Bindlib.box_apply2 (fun a b -> Prod(a,b)) tk b in
+      build_prod k p
+  in
+  let mk_typ = Bindlib.unbox (build_prod k _Type) (*FIXME?*) in
+  let mk = fresh_meta mk_typ k in
+  let tk = _Meta mk (Array.map _Vari vs) in
+  Bindlib.unbox (build_prod k tk)
+
+type rule_ = sym * pp_hint * rule
+
+(** [cp builtins r1 r2] calculates the critical pairs of the rules r1 and r2
+    *)
+let cp : sym StrMap.t -> rule_ -> rule_ -> (term * term) list
+  = fun builtins (s1, h1, r1) (s2, h2, r2) ->
+  let arity1 = Bindlib.mbinder_arity r1.rhs in
+  let arity2 = Bindlib.mbinder_arity r2.rhs in
+  let cps = ref [] in
+  let rec to_m : int -> (meta option) array -> term -> tbox = fun k metas t ->
+    match unfold t with
+    | Vari x         -> _Vari x
+    | Symb (s, h)    -> _Symb s h
+    | Abst (a, t)    ->
+        let (x, t) = Bindlib.unbind t in
+        _Abst (to_m 0 metas a) (Bindlib.bind_var x (to_m 0 metas t))
+    | Appl (t, u)    -> _Appl (to_m (k + 1) metas t) (to_m 0 metas u)
+    | Patt (i, n, a) ->
+        begin
+          let a = Array.map (to_m 0 metas) a in
+          let l = Array.length a in
+          match i with
+          | None   ->
+              let m = fresh_meta ~name:n (build_meta_type (l + k)) l in
+              _Meta m a
+          | Some i ->
+              match metas.(i) with
+              | Some m -> _Meta m a
+              | None   ->
+                  let m = fresh_meta ~name:n (build_meta_type (l + k)) l in
+                  metas.(i) <- Some m;
+                  _Meta m a
+        end
+    | _              -> assert false in
+  let metas1 = Array.init arity1 (fun _ -> None) in
+  let lhs1 = List.map (fun p -> Bindlib.unbox (to_m 0 metas1 p)) r1.lhs in
+  let lhs1 = Basics.add_args (Symb (s1, h1)) lhs1 in
+  let find_cp : term -> unit = fun t ->
+    let metas1 = Array.init arity1 (fun _ -> None) in
+    let metas2 = Array.init arity2 (fun _ -> None) in
+    let lhs2 = List.map (fun p -> Bindlib.unbox (to_m 0 metas2 p)) r2.lhs in
+    let lhs2 = Basics.add_args (Symb (s2, h2)) lhs2 in
+    let fn m =
+      let m = match m with Some m -> m | None -> assert false in
+      let xs = Array.init m.meta_arity (Printf.sprintf "x%i") in
+      let xs = Bindlib.new_mvar mkfree xs in
+      let e = Array.map _Vari xs in
+      TE_Some (Bindlib.unbox (Bindlib.bind_mvar xs (_Meta m e)))
+    in
+    let te_envs1 = Array.map fn metas1 in
+    let te_envs2 = Array.map fn metas2 in
+    let rhs1 = Bindlib.msubst r1.rhs te_envs1 in
+    let rhs2 = Bindlib.msubst r2.rhs te_envs2 in
+    let to_solve = [(t, lhs2)] in
+    match Unif.(solve builtins false {no_problems with to_solve}) with
```

Comment:
As said in a previous comment, finally, this is not a good idea for the moment to use the unification algorithm defined in unif.ml because it is doing unification modulo rewriting. You need to use syntactic unification here. So it is better to write your own unification algorithm.

Reply:
See https://github.com/Deducteam/lambdapi/pull/200#issuecomment-481692245.